### PR TITLE
chore: relax forbidden react server API errors

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

In RSC, components can be shared between server and client, the client components may contain react hooks which will run on the client but shouldn't run on the server. This PR reduces the errors to only throw when APIs are used and not when they're simply imported. The import errors still run when importing `Component` or `PureComponent`.

To fully support shared components, we'll need a follow up PR to "optimize" server components. This should collapse unused hooks so that they don't run, and therefore don't trigger the static error pass. This optimization pass is a bit tricky since we have old babel passes that obfuscate the expected syntax, ref https://github.com/facebook/react-native/pull/43662

# Test Plan

- Updated the tests with the expected syntax.

### CODE REVIEW

File: packages/babel-preset-expo/src/restricted-react-api-plugin.ts [hunk 1]
```diff
@@ -67,10 +67,22 @@ export function environmentRestrictedReactAPIsPlugin(
...
-                );
+                if (['Component', 'PureComponent'].includes(importName)) {
+                  // Add special handling for `Component` since it is different to a function API.
+                  throw path.buildCodeFrameError(
+                    `Client-only "${sourceValue}" API "${importName}" cannot be imported in a React server component. Add the "use client" directive to the top of this file or one of the parent files to enable running this stateful code on a user's device.`
+                  );
+                } else {
+                  const forbiddenImports: Map<string, Set<string>> = path.scope.getData(
+                    'forbiddenImports'
+                  ) || new Map();
```
Comment (ide) [inline packages/babel-preset-expo/src/restricted-react-api-plugin.ts:78]:
```suggestion
                  ) ?? new Map();
```

File: packages/babel-preset-expo/src/restricted-react-api-plugin.ts [hunk 2]
```diff
@@ -83,6 +95,23 @@ export function environmentRestrictedReactAPIsPlugin(
...
+      // Match against `var _useState = useState(0),`
+      VariableDeclarator(path) {
+        const importedSpecifiers: undefined | Map<string, Set<string>> =
+          path.scope.getData('forbiddenImports');
+        if (!importedSpecifiers) return;
+        importedSpecifiers.forEach((forbiddenApis, importName) => {
+          if (t.isCallExpression(path.node.init) && t.isIdentifier(path.node.init.callee)) {
+            if (forbiddenApis.has(path.node.init.callee.name)) {
+              throw path.buildCodeFrameError(
+                `Client-only "useState" API cannot be used in a React server component. Add the "use client" directive to the top of this file or one of the parent files to enable running this stateful code on a user's device.`
```
Comment (ide) [inline packages/babel-preset-expo/src/restricted-react-api-plugin.ts:108]: What is a "parent file"?

Comment (EvanBacon) [reply packages/babel-preset-expo/src/restricted-react-api-plugin.ts:108]: A file that imports this file. The verbiage comes from next swc implementation which is cited in the comments for this plugin https://github.com/vercel/next.js/blob/d43a387d271263f2c1c4da6b9db826e382fc489c/packages/next-swc/crates/next-custom-transforms/src/transforms/react_server_components.rs#L97

Comment (ide) [reply packages/babel-preset-expo/src/restricted-react-api-plugin.ts:108]: OK, I would say more clearly: "one of the files that imports this file". "Parent file" is confusing in the context of a file system where it more likely implies the parent directory.

Comment (ide) [reply packages/babel-preset-expo/src/restricted-react-api-plugin.ts:108]: I still think "parent file" is a misleading term. "dependent module" or the suggestion I wrote earlier would be more accurate and unambiguous.

**Pull request merged**